pkg/internal/authorizer/authorizer.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2019 The KubeCarrier Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authorizer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/gobuffalo/flect"
+	"google.golang.org/grpc/metadata"
+	authv1 "k8s.io/api/authorization/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+const (
+	userHeaderKey = "user"
+)
+
+type AuthorizationClient struct {
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+	client.Client
```

Comment:
don't embed this field; you should implement all `client.Client` function and run authZ layer before proxying the request via the client. 

By embedding the field you're not checking authz for PATCH/UPDATE and some other verbs for example, since you haven't implemented them.

Reply:
Please check the new implementation.

Reply:
Also, I have tried to implement authorizer as a Middleware (Just like our OIDC middleware), register it as `mux.MiddlewareFunc`, so we don't need to handle authorization in the request handler(Get/List, etc.), and we just do the business logic there.
But the issue is that how can I map the endpoint/ query parameters to the resources that the user wants to access? For example, for the offering list endpoint, the HTTP request will like this:
```
GET https://localhost:9443/v1/accounts/{account}/offerings
```
How can I map this to:
User wants to verb: `get`, version: `v1alpha1`, group: `catalog.kubecarrier.io`, resources: `offerings` in namespace: `account`  
Does this mean that I need to build my own `RestMapper` to handle this?
WDYT @nmiculinic @thetechnick @aborilov ?

Reply:
https://github.com/kubermatic/kubecarrier/pull/443/files
here is my proposal for that.
we can use so-called `service wrappers` that isolate business logic from validation/auth/oidc, etc.

Reply:
Well, we can still isolate the business logic without having the wrappers, we can just put the business logic into `handleRequest` functions, like what I did in the PR.

The issue of this Implementation is that we still need to worry about authorization in every endpoint, for example, in my PR, I need to add `Authorize` to the handler when I add new service endpoints, in your approach, we need to add new wrapper for new implemented service endpoints.

If we can implement it like our OIDC middleware, we just register it as a middleware handler, and if new endpoints are added, we don't need to do extra work to make the authorization work.

Reply:
>Well, we can still isolate the business logic without having the wrappers, we can just put the business logic into handleRequest functions, like what I did in the PR.

you already have one `handleRequest` function for `oids` and one for `authorization` and number will keep growing, it's not pleasant to have all of them in one place.

> in your approach, we need to add new wrapper for new implemented service endpoints.

a good thing about this approach, you can easily go to `service_auth` file and find which endpoint has authorization and which one not, in case we want to have endpoint without authorization(like `version`). Same for validation, you don't need to search the handler function where validation start and where business logic ends, you just jump to `service_validation` file. And those files much easier to test, you can test separately authorization and business logic.

>If we can implement it like our OIDC middleware, we just register it as a middleware handler, and if new endpoints are added, we don't need to do extra work to make the authorization work.

that will be the best solution for us. the only problem that we can have, when we want to have some special cases, like leave endpoint without authorization, will be hard to configure that in middleware

Reply:
> you already have one handleRequest function for oids and one for authorization and number will keep growing, it's not pleasant to have all of them in one place.

Nope, I only have one handleRequest for every handler to include business logic, the number will not grow.

Reply:
that mean, you will have put all other logic in one method

Reply:
For account endpoint, we will only have oidc logic to get userInfo, because we need to filter accounts by using the username.
For rest of endpoint, only authorization logic is needed.

Reply:
```
func (o accountServer) List(ctx context.Context, req *v1.AccountListRequest) (res *v1.AccountList, err error) {
	user, present := oidc.ExtractUserInfo(ctx)
	if !present {
		return nil, status.Error(codes.Unauthenticated, "unauthenticated user")
	}
	if err := o.authorizer.Authorize(ctx, &catalogv1alpha1.Account{}, authorizer.AuthorizationOption{
		Namespace: req.Account,
		Verb:      authorizer.RequestList,
	}); err != nil {
		return nil, status.Error(codes.Unauthenticated, err.Error())
	}
	return o.handleListRequest(ctx, req, user.GetName())
```
something like that, right?
this one is hard to test because you have both authorizer and oidc in one place

Reply:
nope, account endpoint will not have authorization logic

Reply:
ok, then

Reply:
This is because we implemented the account endpoint differently, for other endpoints, we use RBAC and SubjectAccessReview to do authorization/permission validation, but for account endpoint, because user can not access accounts anyways(no rbac for them to access accounts), so we don't need to authorize it, we just filter accounts according to accout.Spec.Subject

Reply:
but I still don't like to unit test some `private` method `handleListRequest` instead of the main `List` method

Reply:
> but I still don't like to unit test some `private` method `handleListRequest` instead of the main `List` method

Agree, but the logic we are testing in this PR and your PR are the same right :) ? I mean in this PR, I am testing provider `handleRequest` function, and in your PR, you just test `List` function, and you didn't test wrapper `List` function, right? The number of covered lines should be the same I guess :)

Reply:
correct, but the idea of this wrapper, that if I wrote tests on business logic and business logic doesn't change, I don't need to change tests. In your case, you have to switch unit tests to the private method. but that also ok because you need to do this only once. but then, if you will decide to test authorization, you can do this without call business logic. also, if you add some other logic to `List` method you will again have to fix authorization tests. 
this approach can be very helpful when we will decide to add `audit`, we can simply add new wrapper with audit logic and don't touch any other code

Reply:
>  but then, if you will decide to test authorization, you can do this without call business logic. also, if you add some other logic to List method you will again have to fix authorization tests.
For this one, I don't really know how to unit test authorization stuff, so I would use e2e tests for it, also, if I want to unit test authorization, I will add test to the `authorizer` middleware folder, right? so I don't need to fix authorization tests when I add other logic to `List` method